app/controllers/api/base_controller/authentication.rb

```diff
@@ -2,16 +2,22 @@ module Api
   class BaseController
     module Authentication
       SYSTEM_TOKEN_TTL = 30.seconds
+      SESSION_COOKIE = Vmdb::Application.config.session_options[:key]
 
       def auth_mechanism
         if request.headers[HttpHeaders::MIQ_TOKEN]
           :system
         elsif request.headers[HttpHeaders::AUTH_TOKEN]
           :token
-        elsif request.headers["HTTP_AUTHORIZATION"] && params[:requester_type] == 'ui'
-          :basic_ui
         elsif request.headers["HTTP_AUTHORIZATION"]
           :basic
+        elsif request.cookies[SESSION_COOKIE] || request.x_csrf_token
+          # A successful UI authentication requires both a valid session and a CSRF token,
+          # however, the condition is a little more permissive, because if a session expires
+          # only the CSRF token is being sent and we want the API to know that the request was
+          # initiated from the UI. Otherwise the response to a such request would force the
+          # browser to throw an undesired HTTP basic authentication dialog.
+          :cookie_ui
```

Comment:
What I'm reading in your comment is that the CSRF token will always exist, so why have the check for the session cookie here at all.  That is, it seems this code can just be

```ruby
...
elsif request.x_csrf_token
  :cookie_ui
else
...
```

Reply:
So, that being said, where do we use the SESSION_COOKIE at all after this point?

Reply:
@Fryguy in my tests the session cookie is sent automatically, that's AFAIK default rack behavior. You'll have a session cookie in Rails after your first request to the site, even if you're not logged in. But you're right, I'll simplify this part...